Do not ignore exception on main loop

guibou opened the pull request:
This closes #77.

In the previous implementation, any exception in `runLoop` was ignored
until `Gtk.main` ends. This resulted to weird behaviors where user
`update` / `view` functions where ignored, but the application was still
running and primitive gtk events (such as style change on mouse
overlays) were still running, resulting in an apparently alive but
unresponsive application. The exception was also not printed.

With this change, the application now quit if any exception is raised in
the `view` and `update` user function. User are in charge of setting an
exception handler if they want a different behavior.

Dretch commented:
I think this totally makes sense. I will attempt to write an automated test for this feature, and then I think I have permission to merge the code (might need to fix Travis somehow too).

guibou commented:
@Dretch Thank you for adding a test and taking care of this PR. Do you think you can extend your test with a testcase for an error in the `view` method too?

Dretch commented:
@guibou 
> @Dretch Thank you for adding a test and taking care of this PR. Do you think you can extend your test with a testcase for an error in the view method too?

Good idea. I have pushed a test. It is currently failing, however! I think this is related to the view function being evaluated on the Gtk+ UI thread.

guibou commented:
@Dretch Thank you. Let me update the PR with the fix.

guibou commented:
@Dretch it was a bit more painful than expected.

I addressed `view` and `pipeline` exception. And also, i realized that any exception in `update` deeply inside the transition type may not be caught too.

It is a bit of a mess. I tested by running all the examples.

owickstrom commented:
Good stuff! 👍

Dretch commented on gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:
```diff
@@ -179,6 +183,13 @@ runUI ma = do
+    -- Any exception in the gtk ui thread will be rethrow in the
+    -- calling thread.
+    -- This ensure that this exception won't terminate the application without any control.
+    -- HOWEVER, it may be raised ANYWHERE in the calling thread.
```
The calling thread will always be in `runUI` waiting for `takeMVar` to complete (or _about_ to call `takeMVar`) - right? So there is no question about where the exception will be thrown to - right?

Anyway, this looks good to my (limited) understanding. I suppose an alternative would be to change `runUI` so it catches `SomeException` with `try @SomeException ma >>= putMVar r` and then unwraps the `Either SomeException a` and re-`throwIO`s the `SomeException`. I don't know if that would be better or worse.

guibou replied:
@Dretch my understanding is that `Gdk.threadsAddIdle _ foo` will immediately terminate. It only schedules`foo` to be executed later in the gtk thread. So `runUI_` will terminate immediately. So technically, `runUI`, from my point of view, is a non blocking action and I don't really understand the rational for the `takeMVar` / `putMVar`.

Later, in the gtk main loop, the `IO` action will be called and my raise an exception in the thread which was running `runUI`.

I do admit that I did not really tried to fully understand the logic and perhaps it can be refactored differetly.

Dretch replied:
Hmm, `runUI_` is non-blocking, but `runUI` is blocking. It needs to be because it returns the result of the callback. `takeMVar` / `putMVar` are used to wait for the result to be produced and to pull it from the other thread.

guibou replied:
Haaa, yes. Sorry, you are right. Yes, the logic is perfectly fine and you are right, the exception will be raised inside `runUI`. Sorry.

guibou replied:
(concurrent programming is hard ;)

Dretch commented on gi-gtk-declarative-app-simple/test/Main.hs:
```diff
@@ -0,0 +1,37 @@
+main :: IO ()
+main = hspec $
+  describe "run" $ do
+    it "propagates exceptions from view function" $
```
Looks good :+1: 

I have now thought of another test: what happens if `update'` returns a `Transition s cmd` where `cmd` throws an exception?

guibou replied:
I do think that right now it will be fine because there is no more `forkIO` (so no more ignored exception) or no more `threadsAddIdle`, so no more exception in an uncontrolled thread. But this is worth a test.

Dretch commented on gi-gtk-declarative-app-simple/src/GI/Gtk/Declarative/App/Simple.hs:
```diff
@@ -135,12 +139,12 @@ runLoop App {..} = do
-
-        -- Finally, we loop.
-        loop newState newMarkup events sub newModel
+        snd <$> Async.concurrently
```
I'm no expert on Haskell concurrency, but my naive guess from looking at this is that every `loop` iteration would result in another level being added to an ever-growing stack of in-complete executions of `Async.concurrently` (since none of the executions can finish until `loop` finishes). So this would cause a memory leak.

Is this a valid concern?

guibou replied:
Yes, good point. I totally missed that one. Each iteration will increase the stack.

I need to find another solution. I do really thank you for catching that.

guibou replied:
I think I can use https://hackage.haskell.org/package/async-2.2.2/docs/Control-Concurrent-Async.html#v:link to link the exception with the current thread.

Dretch replied:
:+1: 

Possibly `mapConcurrently` is another option. I.e. `loop` could produce a list that is passed to `mapConcurrently`. Off the top of my head I'm not 100% sure this would be safe from leaks either, though.

guibou commented:
@Dretch I think I fixed the space leak issue. I also added a test for the `Transition` exception you depicted. It is a nightmare because there is 3 places where a transition may raise an exception (`Transition _ undefined`, `Transition _ (pure undefined)`, `Transition _ (pure (Just undefined))`), and the three are not handled the same way ;)

Dretch commented:
Looks great @guibou

guibou commented:
@Dretch with some mistake (removing your commits, now fixed ;)), I rebased on top of `master` with the hope that the new nix environment will help the success of tests.

Dretch commented:
> @Dretch with some mistake (removing your commits, now fixed ;)), I rebased on top of `master` with the hope that the new nix environment will help the success of tests.

I think the `gi-gtk-declarative-app-simple` tests are failing because they run Gtk, which needs a framebuffer and dbus. The `gi-gtk-declarative` tests were already setup with these in Nix, so I have setup the `app-simple` tests the same way (your review is welcome).

guibou commented:
@Dretch looks fine. The code you moved may be improved a bit (for example, I think that `addBuildTool` is more adapted than `addBuildDepend`), but that's orthogonal to this PR.

guibou commented:
And good job to the original author (@owickstrom, I guess) for the wrapper which allows to run a GTK application in a test. I will use that.

Dretch commented:
Thanks @guibou :tada:
